kernel/src/fs/procfs/pid/stat.rs

```diff
@@ -68,36 +73,124 @@ use crate::{
 /// - env_start        : Start address of environment variables.
 /// - env_end          : End address of environment variables.
 /// - exit_code        : Process exit code as returned by waitpid(2).
-pub struct StatFileOps(Arc<Process>);
+pub struct StatFileOps {
+    process_ref: Arc<Process>,
+    thread_ref: Arc<Thread>,
+    /// If `is_pid_stat` is true, this file corresponds to a process-level `/proc/[pid]/stat`.
+    /// Otherwise, this file corresponds to the thread-level `/proc/[pid]/task/[tid]/stat`.
+    is_pid_stat: bool,
+}
 
 impl StatFileOps {
-    pub fn new_inode(process_ref: Arc<Process>, parent: Weak<dyn Inode>) -> Arc<dyn Inode> {
-        ProcFileBuilder::new(Self(process_ref))
-            .parent(parent)
-            .build()
-            .unwrap()
+    pub fn new_inode(
+        process_ref: Arc<Process>,
+        thread_ref: Arc<Thread>,
+        is_pid_stat: bool,
+        parent: Weak<dyn Inode>,
+    ) -> Arc<dyn Inode> {
+        ProcFileBuilder::new(Self {
+            process_ref,
+            thread_ref,
+            is_pid_stat,
+        })
+        .parent(parent)
+        .build()
+        .unwrap()
     }
 }
 
 impl FileOps for StatFileOps {
     fn data(&self) -> Result<Vec<u8>> {
-        let process = &self.0;
+        let process = &self.process_ref;
+        let thread = &self.thread_ref;
+        let posix_thread = thread.as_posix_thread().unwrap();
 
-        let pid = process.pid();
-        let comm = process.executable_path();
-        let state = if process.status().is_zombie() {
-            'Z'
-        } else {
-            'R'
-        };
+        // According to the Linux implementation, a process's `/proc/<pid>/stat` should be
+        // almost identical to its main thread's `/proc/<pid>/task/<pid>/stat`, except for
+        // fields `exit_code`, `wchan`, `min_flt`, `maj_flt`, `gtime`, `utime`, and `stime`.
+        //
+        // Reference: <https://github.com/torvalds/linux/blob/0ff41df1cb268fc69e703a08a57ee14ae967d0ca/fs/proc/array.c#L467-L681>
+
+        let pid = posix_thread.tid();
+        let comm = posix_thread
+            .thread_name()
+            .lock()
+            .as_ref()
+            .and_then(|name| name.as_string())
+            .unwrap_or_else(|| process.executable_path());
+        let state = if thread.is_exited() { 'Z' } else { 'R' };
         let ppid = process.parent().pid();
         let pgrp = process.pgid();
+        let session = process.sid();
+        let tty_nr = process
+            .terminal()
+            .map(|t| t.id().to_linux_dev())
+            .unwrap_or(0);
+        let tpgid = process.terminal().map_or(-1, |_| pgrp as i64);
```

Comment:
Are you sure that `tpgid` is always the same as `pgrp`? I think it should mean the foreground process group for the terminal (which may not be the current process group)?

https://github.com/torvalds/linux/blob/0ff41df1cb268fc69e703a08a57ee14ae967d0ca/fs/proc/array.c#L517-L520

```rust
    let (tty_nr, tpgid) = if let Some(terminal) = process.terminal() {
        (
            terminal.id().to_linux_dev(),
            terminal
                .job_control()
                .foregroud_process()
                .map(|pgrp| pgrp.pgid())
                .unwrap_or(-1),
        )
    } else {
        (0, -1)
    };